File: module_book/src/main/java/com/ebook/book/mvvm/factory/BookViewModelFactory.java
package com.ebook.book.mvvm.factory;

import android.annotation.SuppressLint;
import android.app.Application;

import androidx.annotation.NonNull;
import androidx.annotation.VisibleForTesting;
import androidx.lifecycle.ViewModel;
import androidx.lifecycle.ViewModelProvider;

import com.ebook.book.mvvm.model.BookCommentsModel;
import com.ebook.book.mvvm.model.BookListModel;
import com.ebook.book.mvvm.viewmodel.BookCommentsViewModel;
import com.ebook.book.mvvm.viewmodel.BookListViewModel;

public class BookViewModelFactory implements ViewModelProvider.Factory {
    @SuppressLint("StaticFieldLeak")
    private static volatile BookViewModelFactory INSTANCE;
    private final Application mApplication;

    private BookViewModelFactory(Application application) {
        this.mApplication = application;
    }

    public static BookViewModelFactory getInstance(Application application) {
        if (INSTANCE == null) {
            synchronized (BookViewModelFactory.class) {
                if (INSTANCE == null) {
                    INSTANCE = new BookViewModelFactory(application);
                }
            }
        }
        return INSTANCE;
    }

    @VisibleForTesting
    public static void destroyInstance() {
        INSTANCE = null;
    }

    @NonNull
    @Override
    @SuppressWarnings("unchecked")
    public <T extends ViewModel> T create(@NonNull Class<T> modelClass) {
        if (modelClass.isAssignableFrom(BookListViewModel.class)) {
            return (T) new BookListViewModel(mApplication, new BookListModel(mApplication));
        } else if (modelClass.isAssignableFrom(BookCommentsViewModel.class)) {
            return (T) new BookCommentsViewModel(mApplication, new BookCommentsModel(mApplication));
        }
        throw new IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName());
    }
}

File: module_book/src/main/java/com/ebook/book/mvvm/factory/BookViewModelFactory.kt
package com.ebook.book.mvvm.factory

import androidx.lifecycle.ViewModel
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.viewmodel.CreationExtras
import com.ebook.book.mvvm.model.BookCommentsModel
import com.ebook.book.mvvm.model.BookListModel
import com.ebook.book.mvvm.viewmodel.BookCommentsViewModel
import com.ebook.book.mvvm.viewmodel.BookListViewModel

object BookViewModelFactory : ViewModelProvider.Factory {
    @Suppress("UNCHECKED_CAST")
    override fun <T : ViewModel> create(modelClass: Class<T>, extras: CreationExtras): T {
        // 通过extras获取application
        val mApplication = extras[ViewModelProvider.AndroidViewModelFactory.APPLICATION_KEY]!!
        return when (modelClass) {
            BookListViewModel::class.java -> {
                BookListViewModel(mApplication, BookListModel(mApplication))
            }

            BookCommentsViewModel::class.java -> {
                BookCommentsViewModel(mApplication, BookCommentsModel(mApplication))
            }

            else -> throw IllegalArgumentException("Unknown ViewModel class: " + modelClass.getName())
        } as T
    }
}
